R/SEIRDAge_interventions.R

```diff
@@ -0,0 +1,452 @@
+          # dE <- beta_isolated * S * C %*% I - kappa * E
+          dI <- kappa * E - gamma * I - mu * I
+          dR <- gamma * I
+          dD <- mu * I
+          # dcc <- beta_isolated * S * C %*% I
+          dcc <- S*(beta_isolated * C %*% I * isolated_frac * isIntervention + beta_not_isolated * C %*% (I * (1-isolated_frac) * (1-isIntervention)))
+
+          # return the rate of change
+          list(c(dS, dE, dI, dR, dD, dcc))
+        })
+    }
+
+    
+
+
+    # call ode solver
+    out <- ode(
+      y = state, times = times, func = right_hand_side,
+      parms = parameters, method = solve_method)
+
+    # melt dataframe to wide format
+    output_all <- as.data.frame.array(out)
+
+    out_temp = melt(output_all, 'time')
+
+
+
+    # add compartment and age range columns
+    n_compartment_measurements <- length(times) * age
+    out_temp$compartment = c(replicate(n_compartment_measurements, "S"),
+                             replicate(n_compartment_measurements, "E"),
+                             replicate(n_compartment_measurements, "I"),
+                             replicate(n_compartment_measurements, "R"),
+                             replicate(n_compartment_measurements, "D"),
+                             replicate(n_compartment_measurements, "cc"))
+
+    out_temp$age_range = unlist(rep(object@age_ranges, each=length(times)))
+
+    # drop the old variable column
+    out_temp = out_temp %>%
+      dplyr::select(-.data$variable) %>%
+      dplyr::mutate(compartment=as.factor(.data$compartment)) %>%
+      dplyr::mutate(compartment=forcats::fct_relevel(.data$compartment, "S", "E", "I", "R", "D", "cc")) %>%
+      dplyr::mutate(age_range=as.factor(.data$age_range)) %>%
+      dplyr::mutate(age_range=forcats::fct_relevel(.data$age_range, object@age_ranges))
+
+    # compute incidence and deaths
+    changes <- out_temp %>%
+      dplyr::filter(.data$compartment %in% c("cc", "D")) %>%
+      dplyr::group_by(.data$compartment, .data$age_range) %>%
+      dplyr::mutate(value = c(0, diff(.data$value))) %>%
+      dplyr::mutate(compartment = dplyr::if_else(.data$compartment == "cc", "Incidence",
+                                                 "Deaths")) %>%
+      dplyr::ungroup() %>%
+      as.data.frame()
+
+    # remove cumulative cases column from state vector
+    states = out_temp %>%
+      dplyr::filter(.data$compartment != "cc") %>%
+      droplevels() %>%
+      dplyr::ungroup()
+
+    return(list("states" = states, "changes" = changes))
+  })
+
+
+GetIdxIntervention <- function(object, times){
+  #calculate dt (timestep)
+  dt = times[2] - times[1]
+  Intervention <- rep(0, length(times))
+  n_interventions <- length(object@intervention_parameters[["starts"]])
+
+  for(intv in seq(n_interventions)){
+    Intervention[seq(object@intervention_parameters[["starts"]][intv],
+                       object@intervention_parameters[["stops"]][intv],
+                       dt)] = intv
+  }
+  return(Intervention)
```

Comment:
The current implementation of interventions depends too much on times (which is just where we want to observe the results and should not affect the actual simulation).
The approach taken by the vaccination model https://github.com/Como-DTC-Collaboration/como-models/blob/875e3ac6a6e70d1106dfc569436628f5098d9035/R/SEIRDV.R#L323  may be an improvement

Reply:
(Solved by modifying model structure and setting up a function taking t as variables in dealing with interventions)